server/resp/src/main/java/org/infinispan/server/resp/json/JsonNumOpFunction.java

```diff
@@ -22,27 +23,29 @@
 
 import static java.util.Objects.requireNonNull;
 
-public class JsonNumIncrByFunction
+public class JsonNumOpFunction
       implements SerializableFunction<ReadWriteEntryView<byte[], JsonBucket>, List<Number>> {
    public static final String ERR_PATH_CAN_T_BE_NULL = "path can't be null";
    public static final String ERR_INCREMENT_CANT_BE_NULL = "increment can't be null";
-   public static final AdvancedExternalizer<JsonNumIncrByFunction> EXTERNALIZER = new JsonNumIncrByFunction.Externalizer();
+   public static final AdvancedExternalizer<JsonNumOpFunction> EXTERNALIZER = new JsonNumOpFunction.Externalizer();
 
    byte[] path;
-   byte[] increment;
+   byte[] value;
+   NumOpType numOpType;
```

Comment:
Instead of using an `Enum` to differentiate the type, I think it's better to use polymorphism. That way we're sending one less field over the wire and we don't have to introduce a dedicated enum.

We could have something like:

```java
abstract class AbstractNumOpFunction {

   public List<Number> apply(ReadWriteEntryView<byte[], JsonBucket> entryView) {
     // All common logic
   }

   abstract Number operate(JsonNode numNode, JsonNode valueNode);
}

class JsonNumIncByFunction extends AbstractNumOpFunction {}

class JsonNumMultByFunction extends AbstractNumOpFunction {}
```

This means creating an Externalizer per class, but that will disappear shortly anyway :smiley:

Reply:
we doo this for LEN function as well. do you recommend the update ?

Reply:
Yes I think so. In the LEN case it's even simpler as you can just pass the appropriate `Predicate<JsonNode> condition` to the constructor of the abstract parent class.